docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+  We could also show this for files that were added or removed.
+- Instead of using the slugs on storage (if we decide to use S3), we could use the IDs, that will prevent the need to update the data if the slugs change.
+  But if the slugs change, the builds will probably be different, so we will need to update the data anyway.
+- If a project or version is deleted (or deactivated), we should delete the diff as well.
+
+We could store the changed files sorted by the number of changes, or make that an option in the API,
+or just let the client sort the files as they see fit.
+
+A combination of using the DB and S3 would be really useful,
+as the model will serve as the lock for the API,
+and will be easier to keep track of which build was used to generate the diff
+without having to download the diff file from S3.
+
+API
+---
+
+This operation is expensive, so it won't be available to unauthenticated users.
+And a diff can only be done between versions of the same project that the user has access to.
+
+The endpoint will be:
+
+   GET /api/v3/projects/{project_slug}/versions/{version_slug}/diff/{other_version_slug}/
+
+And the response will be:
+
+.. code:: json
+
+   {
+       "version_a": {"id": 1, "build": {"id": 1}},
+       "version_b": {"id": 2, "build": {"id": 2}},
+       "diff": {
+           "added": [{"file": "new.txt"}],
+           "removed": [{"file": "deleted.txt"}],
+           "modified": [{"file": "changed.txt", "lines": {"added": 1, "removed": 1}}]
+       }
+   }
+
+The version and build can be the full objects, or just the IDs and slugs.
+
+We will generate a lock on this request, to avoid multiple calls to the API for the same versions.
+We can reply with a ``202 Accepted`` if the diff is being calculated in another request.
+
+Integrations
+------------
+
+You may be thinking that once we have an API, it will be just a matter of calling that API from a GitHub action. Wrong!
+
+Doing the API call is easy, but knowing *when* to call it is hard.
+We need to call the API after the build has finished successfully,
+or we will be comparing the files of an incomplete or stale build.
+
+Luckily, we have a webhook that tells us when a build has finished successfully.
+But, we don't want users to have to implement the integration by themselves.
+
+We could:
+
+- Use this as an opportunity to explore using GitHub Apps.
+- Request additional permissions in our existing OAuth2 integration (``project`` scope). Probably not a good idea.
+- Expose this feature in the dashboard for now, and use our GitHub action to simply link to the dashboard.
+  Maybe don't even expose the API to the public, just use it internally.
+- Use a custom `repository dispatch event <https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch>`__
+  to trigger the action from our webhook. This requires the user to do some additional setup,
+  and for our webhooks to support custom headers.
```

Comment:
If we want, we could automate this a little, and just ask the user for a GitHub token, and we create the webhook behind the scenes, and then ask them to use our GitHub action as a final step.

Reply:
I think if we're integrating this feature more, we should build a platform feature for it, and remove the GitHub Action approach. That way we can just have the comment update once the build is successful.

Reply:
That will require using GitHub apps :)